assets/scripts/menus/ShareMenu.jsx

```diff
@@ -0,0 +1,159 @@
+import { getStreet } from '../streets/data_model'
+import { getSharingUrl } from '../util/share_url'
+
+function _getSharingMessage () {
+  let message = ''
+  let street = getStreet()
+
+  if (street.creatorId) {
+    if (isSignedIn() && street.creatorId === getSignInData().userId) {
+      message = `Check out my street, ${street.name}, on Streetmix!`
+    } else {
+      message = `Check out ${street.name} by @${street.creatorId} on Streetmix!`
+    }
+  } else {
+    message = `Check out ${street.name} on Streetmix!`
+  }
+
+  return message
+}
+
+export default class ShareMenu extends React.Component {
+  constructor (props) {
+    super(props)
+
+    this.state = {
+      shareUrl: ''
+    }
+
+    this.update = this.update.bind(this)
+    this.updateLinks = this.updateLinks.bind(this)
+    this.updateSignInPromo = this.updateSignInPromo.bind(this)
+    this.onShow = this.onShow.bind(this)
+  }
+
+  componentDidMount () {
+    this.update()
+  }
+
+  update () {
+    this.updateLinks()
+    this.updateSignInPromo()
+  }
+
+  updateLinks () {
+    const url = getSharingUrl()
+
+    this.setState({ shareUrl: url })
+  }
+
+  updateSignInPromo () {
+    if (!isSignedIn()) {
+      this.signInPromo.classList.add('visible')
+    } else {
+      this.signInPromo.classList.remove('visible')
+    }
+  }
+
+  onShow () {
+    // Make sure links are updated when the menu is opened
+    this.update()
+
+    // Auto-focus and select link when share menu is active
+    window.setTimeout(() => {
+      this.shareViaLinkInput.focus()
+      this.shareViaLinkInput.select()
+    }, 200)
+  }
+
+  onClickShareViaTwitter () {
+    trackEvent('SHARING', 'TWITTER', null, null, false)
+  }
+
+  onClickShareViaFacebook () {
+    trackEvent('SHARING', 'FACEBOOK', null, null, false)
+  }
+
+  onClickSaveAsImage (event) {
+    event.preventDefault()
+    ReactDOM.render(<SaveAsImageDialog />, document.getElementById('dialogs-react'))
+  }
+
+  render () {
+    const shareText = _getSharingMessage()
+    const twitterLink = 'https://twitter.com/intent/tweet' +
+      '?text=' + encodeURIComponent(shareText) +
+      '&url=' + encodeURIComponent(this.state.shareUrl)
+
+    const facebookLink = 'https://www.facebook.com/dialog/feed' +
+      '?app_id=' + encodeURIComponent(FACEBOOK_APP_ID) +
+      '&redirect_uri=' + encodeURIComponent(this.state.shareUrl) +
+      '&link=' + encodeURIComponent(this.state.shareUrl) +
+      '&name=' + encodeURIComponent(getPageTitle()) +
+      '&description=' + encodeURIComponent(shareText)
+
+    return (
+      <Menu name='share' alignment='right' onShow={this.onShow} className='share-menu' {...this.props}>
+        <div
+          className='share-sign-in-promo'
+          ref={(ref) => { this.signInPromo = ref }}
+          dangerouslySetInnerHTML={{ __html: t('menu.share.sign-in', '<a href="/twitter-sign-in?redirectUri=/just-signed-in">Sign in with Twitter</a> for nicer links to your streets and your personal street gallery') }}
```

Comment:
It's not a guarantee that in other languages the sentence structure will always be "link + text". It might be reversed, as in "text + link", or even inline, "text + link + text." I don't know what the best way to do it is, but my concern about splitting it up is that it's harder to translate because you can't see the whole sentence.